src/embed/dispatch.rs

```diff
@@ -27,272 +22,40 @@ lazy_static! {
-    }
-}
-
-async fn execute<T, E, F>(func: F, txns: &mut HashMap<u32, db::Write<'_>>, req: Request)
-where
-    T: DeJson + TransactionRequest,
-    E: std::fmt::Debug,
-    F: for<'r, 's> AsyncFn2<&'r mut db::Write<'s>, T, Output = Result<String, E>>,
-{
-    let request: T = match DeJson::deserialize_json(&req.data) {
-        Ok(v) => v,
-        Err(e) => {
-            req.response.send(Err(format!("InvalidJson({})", e))).await;
-            return;
-        }
-    };
-
-    let txn_id = request.transaction_id();
-    let mut txn = match txns.get_mut(&txn_id) {
-        Some(v) => v,
-        None => {
-            req.response
-                .send(Err(format!("No such transaction {}", txn_id)))
-                .await;
-            return;
-        }
-    };
-
-    let response = func
-        .call(&mut txn, request)
-        .await
-        .map_err(|e| format!("{:?}", e));
-    req.response.send(response).await;
-}
-
-trait TransactionRequest {
-    fn transaction_id(&self) -> u32;
-}
-
-#[derive(SerJson)]
-struct OpenTransactionResponse {
-    #[nserde(rename = "transactionId")]
-    transaction_id: u32,
-}
-
-#[derive(DeJson)]
-struct GetRequest {
-    #[nserde(rename = "transactionId")]
-    transaction_id: u32,
-    key: String,
-}
-
-impl TransactionRequest for GetRequest {
-    fn transaction_id(&self) -> u32 {
-        return self.transaction_id;
-    }
-}
-
-#[derive(SerJson)]
-struct GetResponse {
-    value: Option<String>,
-    has: bool, // Second to avoid trailing comma if value == None.
-}
-
-#[derive(DeJson)]
-struct PutRequest {
-    #[nserde(rename = "transactionId")]
-    transaction_id: u32,
-    key: String,
-    value: String,
-}
-
-impl TransactionRequest for PutRequest {
-    fn transaction_id(&self) -> u32 {
-        return self.transaction_id;
-    }
-}
-
-struct Dispatcher {
-    connections: HashMap<String, Sender<Request>>,
-}
-
-impl Dispatcher {
-    async fn open(&mut self, req: &Request) -> Response {
-        if req.db_name.is_empty() {
-            return Err("db_name must be non-empty".into());
-        }
-        if self.connections.contains_key(&req.db_name[..]) {
-            return Ok("".into());
-        }
-        match IdbStore::new(&req.db_name[..]).await {
-            Err(e) => {
-                return Err(format!("Failed to open \"{}\": {}", req.db_name, e));
-            }
-            Ok(v) => {
-                if let Some(kv) = v {
-                    let (tx, rx) = channel::<Request>(1);
-                    spawn_local(connection_loop(dag::Store::new(Box::new(kv)), rx));
-                    self.connections.insert(req.db_name.clone(), tx);
-                }
+        let req = match rx.recv().await {
+            Ok(req) => req,
+            Err(why) => {
+                warn!("Dispatch loop recv failed: {}", why);
+                continue;
             }
+        };
+
+        let response = match req.rpc.as_str() {
+            "open" => Some(do_open(&mut conns, &req).await),
```

Comment:
Wow, this one surprised me. So we can use `rep.rpc.as_str()` or `&*req.rpc` or ` &req.rpc[..]`, but not `req.rpc.into()` for this conversion from `String` to `&str`.

I guess I understand why Rust has these two string types, but I don't understand why comparing a typed variable and a constant value was never made easier.

```
   Compiling replicache-client v0.1.0 (/Users/nate/roci/repc)
error[E0277]: the trait bound `&str: std::convert::From<std::string::String>` is not satisfied
  --> src/embed/connection.rs:43:19
   |
43 |     match req.rpc.into() {
   |                   ^^^^ the trait `std::convert::From<std::string::String>` is not implemented for `&str`
   |
   = note: to coerce a `std::string::String` into a `&str`, use `&*` as a prefix
   = note: `std::convert::From<std::string::String>` is implemented for `&mut str`, but not for `&str`
   = note: required because of the requirements on the impl of `std::convert::Into<&str>` for `std::string::String`
```